packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2150,12 +2183,26 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
      */
     public submitSignal(type: string, content: any) {
         this.verifyNotClosed();
-        const envelope: ISignalEnvelope = { address: undefined, contents: { type, content } };
+        const envelope: ISignalEnvelope = {
+            address: undefined,
+            contents: { type, content },
+            metadata: {
+                sequenceNumber: ++this._perfSignalData.signalSequenceNumber,
+                timestamp: Date.now(),
+            },
+        };
         return this.context.submitSignalFn(envelope);
     }
 
     public submitDataStoreSignal(address: string, type: string, content: any) {
-        const envelope: ISignalEnvelope = { address, contents: { type, content } };
+        const envelope: ISignalEnvelope = {
+            address,
+            contents: { type, content },
+            metadata: {
+                sequenceNumber: ++this._perfSignalData.signalSequenceNumber,
+                timestamp: Date.now(),
```

Comment:
I'd prefer to have same logic as we do with ops - decide what signal to track (based on its client sequence number), remember locally time when we produced that signal, and do the math of receiving signal back from ordering service. All tracking is in memory, with no need to mark ops / signals in any way - all we need is unique identification of ops / signals through client sequence number.